### PR TITLE
feat: add Interim Care Order (FPLA-2948)

### PR DESCRIPTION
**Pull request by jacmurphy**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-2948

### Change description ###

As an admin or judge
I want to create interim care orders
So that I can maintain existing functionality

https://user-images.githubusercontent.com/4429628/119831152-e5b3de00-bef4-11eb-9fe5-5c1f898b0463.mov

[c33_interim_care_order.pdf](https://github.com/hmcts/fpl-ccd-configuration/files/6554171/c33_interim_care_order.pdf)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            errors.add(END_DATE_MAX_RANGE_MESSAGE);
+        }
+
+        return errors;
+    }
+
+    private List<String> validateEndOfProceedings() {
+        List<String> errors = new ArrayList<>();
+
+        log.info("No validation needed for SET_END_OF_PROCEEDINGS");
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java:89]: Needed?

File: ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json [hunk 2]
```diff
@@ -100,7 +100,7 @@
     "PageID": "OrderDetails",
     "PageDisplayOrder": 5,
     "PageColumnNumber": 1,
-    "FieldShowCondition": "orderTempQuestions.epoTypeAndPreventRemoval=\"YES\" AND manageOrdersEpoType=\"PREVENT_REMOVAL\" AND manageOrdersExclusionRequirement=\"Yes\"",
+    "FieldShowCondition": "orderTempQuestions.epoTypeAndPreventRemoval=\"YES\" AND manageOrdersEpoType=\"PREVENT_REMOVAL\"",
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply ccd-definition/CaseEventToFields/CareSupervision/manageOrders/5_orderDetailsSection.json:103]: Youre correct, .this is a mistake on my part thank you for the spot. Removing now

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        final LocalDateTime endDateTime = manageOrdersEventData.getManageOrdersSetDateAndTimeEndDate();
+        final LocalDate approvalDate = manageOrdersEventData.getManageOrdersApprovalDate();
+
+        switch (type) {
+            case SET_CALENDAR_DAY:
+                return validateDate(approvalDate, endDate);
+            case SET_CALENDAR_DAY_AND_TIME:
+                return validateDateTime(approvalDate, endDateTime);
+            case SET_END_OF_PROCEEDINGS:
+                return validateEndOfProceedings();
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java:50]: This does return an empty array, however it keeps implimentation consistent and logs that there is no validation needed for this selection. I think it adds more value in this way

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C33InterimCareOrderDocumentParameterGenerator.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    @Override
+    public DocmosisTemplates template() {
+        return DocmosisTemplates.ORDER;
+    }
+
+    private String orderDetails(int numOfChildren, String localAuthorityName, ManageOrdersEventData eventData) {
+        LocalDateTime orderExpiration;
+        String formatString;
+        String courtResponsibilityAssignmentMessage = "The Court orders %s supervises the %s until %s.";
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/generator/C33InterimCareOrderDocumentParameterGenerator.java:71]: I initially had it as simple as this but once i noticed the word document is quite difficult to read as a template, i wanted to make it very clear which section this message was refering to.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java (hunk 1 above)
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java:89]: Spoke about this offline - My point was that it is good for developers going forward to understand WHY this is empty

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/Order.java [hunk 5]
```diff
@@ -47,7 +49,20 @@
             APPROVAL_DATE,
             WHICH_CHILDREN,
             FURTHER_DIRECTIONS,
-            SUPERVISION_ORDER_END_DATE,
+            MANAGE_ORDER_END_DATE_WITH_MONTH,
+            REVIEW_DRAFT_ORDER)
+    ),
+    C33_INTERIM_CARE_ORDER(
+        "Interim care order",
+        "Section 44 Children Act 1989",
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/order/Order.java:57]: Just got conformation that its 38. Updating now

File: e2e/pages/events/manageOrders.page.js [hunk 6]
```diff
@@ -195,11 +230,15 @@ const checkPreview = async () => {
   await I.runAccessibilityTest();
 };
 
+const selectExclusionRequirementICO = async (exclusionRequirement) => {
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply e2e/pages/events/manageOrders.page.js:233]: removing

File: e2e/pages/events/manageOrders.page.js [hunk 7]
```diff
@@ -181,8 +212,12 @@ const enterSuperVisionNumOfMonths = async (months) => {
   I.fillField(section4.supervisionOrderNumOfMonths, months);
 };
 
-const selectSupervisionOrder = async (orderDateType) => {
-  I.click(`${section4.supervisionOrderType.group}-${orderDateType}`);
+const selectOrderTypeWithMonth = async (orderDateType) => {
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply e2e/pages/events/manageOrders.page.js:215]: removing

File: e2e/pages/events/manageOrders.page.js [hunk 8]
```diff
@@ -181,8 +212,12 @@ const enterSuperVisionNumOfMonths = async (months) => {
...
+  I.click(`${section4.orderTypeWithMonth.group}-${orderDateType}`);
+};
+
+const selectOrderTypeWithEndOfProceedings = async (orderDateType) => {
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply e2e/pages/events/manageOrders.page.js:219]: removing

File: ccd-definition/FixedLists/CareSupervision/ManageOrdersEndDateTypeWithEndOfProceedings.json [hunk 9]
```diff
@@ -0,0 +1,23 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "ManageOrdersEndDateTypeWithEndOfProceedings",
+    "ListElementCode": "SET_CALENDAR_DAY",
```
[the remark this answers is not in the document]
Comment (jacmurphy) [reply ccd-definition/FixedLists/CareSupervision/ManageOrdersEndDateTypeWithEndOfProceedings.json]: Great idea. Updated now

File: e2e/tests/adminManagesOrders_test.js [hunk 10]
```diff
@@ -132,19 +132,19 @@ Scenario('Create C21 blank order', async ({I, caseViewPage, manageOrdersEventPag
...
   await I.goToNextPage();
   await manageOrdersEventPage.enterJudge();
   await I.goToNextPage();
   await manageOrdersEventPage.selectChildren(manageOrdersEventPage.section3.allChildren.options.select, [0]);
   await I.goToNextPage();
   await manageOrdersEventPage.enterFurtherDirections('Supervision order further details.');
-  await manageOrdersEventPage.selectSupervisionOrder(manageOrdersEventPage.section4.supervisionOrderType.options.numberOfMonths);
-  await manageOrdersEventPage.enterSuperVisionNumOfMonths(12);
+  await manageOrdersEventPage.selectOrderTypeWithMonth(manageOrdersEventPage.section4.orderTypeWithMonth.options.numberOfMonths);
+  await manageOrdersEventPage.enterSuperVisionNumOfMonths(10);
```
Comment (samuelzerozero) [inline e2e/tests/adminManagesOrders_test.js:147]: keep it boundary to 12 Months limit is a bit better?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java (hunk 1 above)
Comment (samuelzerozero) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidator.java:89]: I'd remove this log line or up the level of logging to Trace/Debug. Utterly verbose logs are difficult to navigate, especially in production. If you wanted to make it clearer for the developer, you could associate to the empty array a constant that is NO_VALIDATION = Collections.emptyList() or something similar.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+
+    private final ManageOrderEndDateWithEndOfProceedingsValidator underTest =
+        new ManageOrderEndDateWithEndOfProceedingsValidator(time);
+
+    @Test
+    void accept() {
+        assertThat(underTest.accept()).isEqualTo(MANAGE_ORDER_END_DATE_WITH_END_OF_PROCEEDINGS);
+    }
+
+    // Date
```
Comment (samuelzerozero) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/orders/validator/ManageOrderEndDateWithEndOfProceedingsValidatorTest.java:41]: you could remove comment and if you want to group them, you can use @Nested test class

**Review by samuelzerozero — COMMENTED** [review]
Looks good apart from a couple of minors. Question around the name of the exclusionDetails in docmosi template, similar to exclusion clause

**Review by samuelzerozero — APPROVED** [review]

**Review by samuelzerozero — APPROVED** [review]

**Pull request merged**